### ISSUE TITLE
(With fix) A minor mistake in the OpenSL ES wrapper.

### ISSUE THREAD

**Issue by serhii-k** [google/oboe#624]
First of all thank you very much for such a great library!

I've tried to move to Oboe, but have noticed an issue (also there is a fix). In my app I need to pause/resume the output stream.

Let's add 2 buttons ("pause" and "resume") to the hello-oboe sample:

```
    <Button
        android:id="@+id/pauseButton"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintTop_toBottomOf="@id/userInstructionView"
        android:layout_marginTop="16dp"
        android:text="PAUSE" />

    <Button
        android:id="@+id/resumeButton"
        android:layout_width="match_parent"
        android:layout_height="wrap_content"
        app:layout_constraintStart_toStartOf="parent"
        app:layout_constraintEnd_toEndOf="parent"
        app:layout_constraintTop_toBottomOf="@id/pauseButton"
        android:layout_marginTop="16dp"
        android:text="RESUME" />
```

In MainActivity.java:
```
        findViewById(R.id.pauseButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                PlaybackEngine.pause();
            }
        });

        findViewById(R.id.resumeButton).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                PlaybackEngine.resume();
            }
        });
```

PlaybackEngine.java
```
    static void pause() {
        if (mEngineHandle != 0) native_pauseEngine(mEngineHandle);
    }

    static void resume() {
        if (mEngineHandle != 0) native_resumeEngine(mEngineHandle);
    }

    private static native void native_pauseEngine(long engineHandle);
    private static native void native_resumeEngine(long engineHandle);
```

HelloOboeEngine.h
```
    void pause();

    void resume();
```

HelloOboeEngine.cpp
```
void HelloOboeEngine::pause() {
    mStream->requestPause();
}

void HelloOboeEngine::resume() {
    mStream->requestStart();
}
```

jni-bridge.cpp
```
JNIEXPORT void JNICALL
Java_com_google_sample_oboe_hellooboe_PlaybackEngine_native_1pauseEngine(JNIEnv *env, jclass,
                                                                         jlong engineHandle) {
    HelloOboeEngine *engine = reinterpret_cast<HelloOboeEngine*>(engineHandle);
    if (engine == nullptr) {
        LOGE("Engine is null, you must call createEngine before calling this method");
        return;
    }
    engine->pause();
}

JNIEXPORT void JNICALL
Java_com_google_sample_oboe_hellooboe_PlaybackEngine_native_1resumeEngine(JNIEnv *env, jclass,
                                                                          jlong engineHandle) {
    HelloOboeEngine *engine = reinterpret_cast<HelloOboeEngine*>(engineHandle);
    if (engine == nullptr) {
        LOGE("Engine is null, you must call createEngine before calling this method");
        return;
    }
    engine->resume();
}
```

Everything works as expected with AAudio, but with OpenSL ES it works this way:
1) At the beginning the audio stream is started.
2) Press "pause", then "resume". The stream still plays.
3) Press "pause", then "resume" again. Here the stream goes to the "Stopped" state! No sound when tapped. Unexpectedly...
4) After another "pause" and "resume" the stream plays again.
... The stream then goes to the "Stopped" state on each 3rd "resume".

So, this issue happens when resuming the output stream. On each 3rd requestStart(). { In fact on each kBufferQueueLength+1 resume }

Android log:
D/OboeAudio: AudioOutputStreamOpenSLES(): requestStart() called
D/OboeAudio: AudioOutputStreamOpenSLES(): setPlayState_l() called
D/OboeAudio: AudioOutputStreamOpenSLES(): requestStart() returning 0
W/libOpenSLES: Leaving BufferQueue::Enqueue (SL_RESULT_BUFFER_INSUFFICIENT)
E/OboeAudio: enqueueCallbackBuffer() returned 7
D/OboeAudio: AudioOutputStreamOpenSLES(): requestStop() called
D/OboeAudio: AudioOutputStreamOpenSLES(): setPlayState_l() called
D/OboeAudio: AudioOutputStreamOpenSLES(): requestFlush_l() called
D/OboeAudio: AudioOutputStreamOpenSLES(): requestFlush_l() returning 0
D/OboeAudio: AudioOutputStreamOpenSLES(): requestStop() returning 0

So after Enqueue returns SL_RESULT_BUFFER_INSUFFICIENT (code 7), Oboe intentionally stops the stream:
https://github.com/google/oboe/blob/d153035cf7db0c822c4246c47feffe05a103eb14/src/opensles/AudioStreamOpenSLES.cpp#L253

///
**What makes Enqueue return SL_RESULT_BUFFER_INSUFFICIENT?**

From the OpenSL ES specs:
> If the maximum number of buffers specified in
> the SLDataLocator_BufferQueue structure used as the data source
> when creating the media object using the CreateAudioPlayer or
> CreateMidiPlayer method has been reached, the buffer is not added
> to the buffer queue and SL_RESULT_BUFFER_INSUFFICIENT is
> returned.

In Oboe we have kBufferQueueLength=2 buffers. The issue happens if we call Enqueue 3 times.

This happens in AudioOutputStreamOpenSLES::requestStart:
https://github.com/google/oboe/blob/449c084b6bc62872e6cdcbd3cc7afcd403de54a3/src/opensles/AudioOutputStreamOpenSLES.cpp#L300

Basically, processBufferCallback() should be called only when the stream has not yet been ever started. We only enqueue the first buffer.

My quick fix was:
```
if (initialState == StreamState::Open) {
    processBufferCallback(mSimpleBufferQueueInterface);
}
```

But, I'm not sure. Maybe some other states are also appropriate here?

**Comment by philburk** [comment]
Thanks for the report. Please let us know your
Android Version(s): ??
Android Device(s): ??
Oboe Version: ??

I tried to reproduce this with OboeTester on Q on a Pixel 3A.
Click TEST OUTPUT
Select API: OpenSL ES
Click OPEN
Click START
Hear sound
Click PAUSE
Click START
Hear sound
Click PAUSE
Click START
Hear sound
Click PAUSE
and so on....

So I could not reproduce the bug. How does OboeTester behave on your system?

https://github.com/google/oboe/tree/master/apps/OboeTester/docs

I notice that in your log you got:
    D/OboeAudio: AudioOutputStreamOpenSLES(): requestStop() called
I got:
    AudioOutputStreamOpenSLES(): requestPause() called

**Comment by serhii-k** [comment]
@philburk Thank you for your reply.

Oboe version: the latest cloned from the repo (the master branch).
Android version(s) and device(s):
8.0 on Huawei Honor 8
8.1 on Android Emulator
But I'm pretty sure, this issue is replicable on any Android device/emulator when using the OpenSL ES implementation.

I have tried the OboeTester. There is no such issue when starting the stream after the pause.
But... After looking on its implementation, the "Start" button actually Stops the stream before requesting it to Start.
So, in OboeTester we get the following calls:
- after pressing "Pause" => requestPause()
- after pressing "Start" then => requestStop(), requestStart()

Calling the intermediate requestStop clears the OpenSL ES buffers, so this issue doesn't happen. But this call is unnecessary and it introduces an audio glitch during pause-stop-start.

===

Please try to add the proposed buttons to the hello-oboe example. I'm sure you will get the described behavior on any device. There is just a minor mistake in the Oboe implementation of the OpenSL ES wrapper (it does excessive calls to Enqueue, and overfills the OpenSL ES buffers).

> I notice that in your log you got
Those lines are logged after pressing the "Resume" button, just before the stream is stopped by Oboe (when getting SL_RESULT_BUFFER_INSUFFICIENT).

Sorry if I wrote unclear. Please let me know if I should explain something in more details. Thank you!

**Comment by philburk** [comment]
> [in Oboetester] the "Start" button actually Stops the stream before requesting it to Start.

Thank you so much for this analysis. And the additional info. It was very helpful. I removed the call to stop() from the start code and was able to reproduce the bug.

The problem was that I was adding a buffer to a queue that was already full. I only needed to add when the queue was empty to start the stream. So for the fix I added a method to query the queue depth and only added the buffer if depth is zero.
